### PR TITLE
port most simple headers to cats parse

### PR DESCRIPTION
**Pull request by m-sp**
well that blew up more than intended 🙈
hope this is still reviewable, let me know if I should split this up into smaller chunks

### CODE REVIEW

File: core/src/main/scala/org/http4s/headers/Host.scala [hunk 1]
```diff
@@ -17,14 +17,31 @@
...
   override def parse(s: String): ParseResult[Host] =
-    HttpHeaderParser.HOST(s)
+    ParseResult.fromParser(parser, "Invalid Host")(s)
+
+  //  // Do not accept scoped IPv6 addresses as they should not appear in the Host header,
+  //  // see also https://issues.apache.org/bugzilla/show_bug.cgi?id=35122 (WONTFIX in Apache 2 issue) and
+  //  // https://bugzilla.mozilla.org/show_bug.cgi?id=464162 (FIXED in mozilla)
+  private[http4s] val parser = {
+    val port = Parser.string(":") *> Rfc3986.digit.rep.string.mapFilter { case s =>
+      try Some(s.toInt)
```
Comment (hamnis) [inline core/src/main/scala/org/http4s/headers/Host.scala:36]: Try{ s.toInt }.toOption feels like it reads better.

File: core/src/main/scala/org/http4s/headers/Host.scala [hunk 2]
```diff
@@ -17,14 +17,31 @@
...
+      catch { case _: NumberFormatException => None }
+    }
+
+    val host = Rfc2616.token.orElse(Uri.Host.ipLiteral.string).backtrack <* Rfc2616BasicRules.optWs
```
Comment (hamnis) [inline core/src/main/scala/org/http4s/headers/Host.scala:40]: We have two versions of this token and whitespace rules now.
I  suggest we try to use `Rfc7230.token` and `Rfc7230.ows` where possible

File: core/src/main/scala/org/http4s/headers/Server.scala [hunk 3]
```diff
@@ -17,15 +17,23 @@
...
 
 object Server extends HeaderKey.Internal[Server] with HeaderKey.Singleton {
   def apply(id: ProductId): Server =
     new Server(id, Nil)
 
   override def parse(s: String): ParseResult[Server] =
-    HttpHeaderParser.SERVER(s)
+    ParseResult.fromParser(parser, "invalid Server")(s)
+
+  private[http4s] val parser = {
```
Comment (hamnis) [inline core/src/main/scala/org/http4s/headers/Server.scala:30]: `User-Agent` and Server are very similar, maybe they should share the parser up to the last map?

**Comment by hamnis** [comment]
Well done! You beat me to it =)

File: core/src/main/scala/org/http4s/headers/Allow.scala [hunk 4]
```diff
@@ -17,14 +17,20 @@
...
+import org.http4s.internal.parsing.Rfc7230
 import org.http4s.util.Writer
 
 object Allow extends HeaderKey.Internal[Allow] with HeaderKey.Singleton {
   def apply(ms: Method*): Allow = Allow(ms.toSet)
 
   override def parse(s: String): ParseResult[Allow] =
-    HttpHeaderParser.ALLOW(s)
+    ParseResult.fromParser(parser, "Invalid Allow")(s)
+  private[http4s] val parser = Rfc7230
```
Comment (hamnis) [inline core/src/main/scala/org/http4s/headers/Allow.scala:28]: alternative version: `  private[http4s] val parser = Method.parser
    .repSep0(Parser.char(',').surroundedBy(Rfc7230.ows))
    .map(list => Allow(list.toSet))`

Comment (m-sp) [reply core/src/main/scala/org/http4s/headers/Allow.scala:28]: hmm not sure if this actually equivivalent

https://tools.ietf.org/html/rfc7230#section-7

allows e.g. `"foo , ,bar,charlie   "` as a valid value for the `#element` syntax

the header is just weirdly specified https://tools.ietf.org/html/rfc7231#section-7.4.1
the syntax is described as `Allow = #method` which would mean at least one method, but they immediately contradict the syntax: `An empty Allow field value indicates that the resource allows no methods`

File: core/src/main/scala/org/http4s/headers/Server.scala (hunk 3 above)
Comment (m-sp) [reply core/src/main/scala/org/http4s/headers/Server.scala]: moved them to the `AdditionalParsers` if you have a better location in mind please let me know
https://github.com/http4s/http4s/pull/4208/commits/b89f9d6ebfbea75f7c51749d262d6361146b3d44

File: core/src/main/scala/org/http4s/headers/Host.scala (hunk 2 above)
Comment (m-sp) [reply core/src/main/scala/org/http4s/headers/Host.scala]: sounds good https://github.com/http4s/http4s/pull/4208/commits/b89f9d6ebfbea75f7c51749d262d6361146b3d44

File: core/src/main/scala/org/http4s/headers/Host.scala (hunk 1 above)
Comment (m-sp) [reply core/src/main/scala/org/http4s/headers/Host.scala]: fine for me https://github.com/http4s/http4s/pull/4208/commits/b89f9d6ebfbea75f7c51749d262d6361146b3d44

File: core/src/main/scala/org/http4s/headers/Allow.scala (hunk 4 above)
Comment (hamnis) [reply core/src/main/scala/org/http4s/headers/Allow.scala:28]: hm, interesting. maybe we need a test that also adds that edge-case.

File: core/src/main/scala/org/http4s/ProductIdOrComment.scala [hunk 5]
```diff
@@ -32,9 +34,19 @@ final case class ProductId(value: String, version: Option[String] = None)
   }
 }
 
+object ProductId {
+  private[http4s] val parser = (Rfc2616.token ~ (Parser.string("/") *> Rfc2616.token).?).map {
```
Comment (hamnis) [inline core/src/main/scala/org/http4s/ProductIdOrComment.scala:38]: replace with Rfc7230.token

File: core/src/main/scala/org/http4s/headers/Host.scala [hunk 6]
```diff
@@ -17,14 +17,30 @@
...
+
+  //  // Do not accept scoped IPv6 addresses as they should not appear in the Host header,
+  //  // see also https://issues.apache.org/bugzilla/show_bug.cgi?id=35122 (WONTFIX in Apache 2 issue) and
+  //  // https://bugzilla.mozilla.org/show_bug.cgi?id=464162 (FIXED in mozilla)
+  private[http4s] val parser = {
+    val port = Parser.string(":") *> Rfc3986.digit.rep.string.mapFilter { s =>
+      Try(s.toInt).toOption
+    }
+
+    val host = Rfc7230.token.orElse(Uri.Host.ipLiteral.string).backtrack <* Rfc7230.ows
```
Comment (hamnis) [inline core/src/main/scala/org/http4s/headers/Host.scala:39]: not sure if this is correct.

Ref comment above, Ipv6 should be disallowed, and this does not include ipv4 ?

Comment (hamnis) [reply core/src/main/scala/org/http4s/headers/Host.scala:39]:
``` private[http4s] val parser = {
    val host =
      (Uri.Ipv4Address.parser).orElse(Uri.RegName.parser)
    val port = Parser.string(":") *> Rfc3986.digit.rep.string.mapFilter { s =>
      Try(s.toInt).toOption
    }

    (host ~ port.?).map { case (host, port) =>
      Host(host.value, port)
    }
  }```
Possible implementation

Comment (m-sp) [reply core/src/main/scala/org/http4s/headers/Host.scala:39]: Hmm you are absolutely right. Weird thing is that I'm pretty sure this is than wrong in 0.21

Comment (m-sp) [reply core/src/main/scala/org/http4s/headers/Host.scala]: https://github.com/http4s/http4s/pull/4208/commits/69051b341fba89b474cb27fa12f83755758f458f

File: core/src/main/scala/org/http4s/ProductIdOrComment.scala (hunk 5 above)
Comment (m-sp) [reply core/src/main/scala/org/http4s/ProductIdOrComment.scala]: https://github.com/http4s/http4s/pull/4208/commits/8663fa733c3c613c41ea5b37efacabce4c1e66e8

File: core/src/main/scala/org/http4s/headers/Host.scala [hunk 7]
```diff
@@ -17,14 +17,31 @@
...
 object Host extends HeaderKey.Internal[Host] with HeaderKey.Singleton {
   def apply(host: String, port: Int): Host = apply(host, Some(port))
 
   override def parse(s: String): ParseResult[Host] =
-    HttpHeaderParser.HOST(s)
+    ParseResult.fromParser(parser, "Invalid Host")(s)
+
+  //  // Do not accept scoped IPv6 addresses as they should not appear in the Host header,
+  //  // see also https://issues.apache.org/bugzilla/show_bug.cgi?id=35122 (WONTFIX in Apache 2 issue) and
+  //  // https://bugzilla.mozilla.org/show_bug.cgi?id=464162 (FIXED in mozilla)
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/Host.scala:33]: This was a rule in RFC2616 (which referred to the RFC2396 URI, and predates the current URI).  It's allowed in RFC7230.  I think those bugzillas issues date back to the older spec.

File: core/src/main/scala/org/http4s/headers/Host.scala [hunk 8]
```diff
@@ -17,14 +17,31 @@
...
+  private[http4s] val parser = {
+    val host =
+      (Uri.Ipv4Address.parser).orElse(Uri.RegName.parser)
+    val port = Parser.string(":") *> Rfc3986.digit.rep.string.mapFilter { s =>
+      Try(s.toInt).toOption
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/Host.scala:38]: A horrid thing about this spec is that the port can be unbounded.  I've been thinking about this in the URI redesign, as well.  I don't think it's _useful_, but it stands in the way of compliance.

I think this is fine for now, but I'm trying to think of a way out that's both practical and pedantic.

File: core/src/main/scala/org/http4s/headers/If-Match.scala [hunk 9]
```diff
@@ -31,7 +32,14 @@ object `If-Match` extends HeaderKey.Internal[`If-Match`] with HeaderKey.Singleto
     `If-Match`(Some(NonEmptyList.of(first, rest: _*)))
 
   override def parse(s: String): ParseResult[`If-Match`] =
-    HttpHeaderParser.IF_MATCH(s)
+    ParseResult.fromParser(parser, "Invalid If-Match")(s)
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/If-Match.scala:35]:
```suggestion
    ParseResult.fromParser(parser, "Invalid If-Match header")(s)
```

File: core/src/main/scala/org/http4s/headers/If-None-Match.scala [hunk 10]
```diff
@@ -41,7 +42,14 @@ object `If-None-Match` extends HeaderKey.Internal[`If-None-Match`] with HeaderKe
     `If-None-Match`(Some(NonEmptyList.of(first, rest: _*)))
 
   override def parse(s: String): ParseResult[`If-None-Match`] =
-    HttpHeaderParser.IF_NONE_MATCH(s)
+    ParseResult.fromParser(parser, "Invalid If-None-Match")(s)
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/If-None-Match.scala:45]:
```suggestion
    ParseResult.fromParser(parser, "Invalid If-None-Match header")(s)
```

File: core/src/main/scala/org/http4s/headers/Max-Forwards.scala [hunk 11]
```diff
@@ -45,6 +45,8 @@ object `Max-Forwards` extends HeaderKey.Internal[`Max-Forwards`] with HeaderKey.
     fromLong(length).fold(throw _, identity)
 
   override def parse(s: String): ParseResult[`Max-Forwards`] =
-    HttpHeaderParser.MAX_FORWARDS(s)
+    ParseResult.fromParser(parser, "invalid Max-Forwards")(s)
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/Max-Forwards.scala:48]:
```suggestion
    ParseResult.fromParser(parser, "invalid Max-Forwards header")(s)
```

File: core/src/main/scala/org/http4s/headers/Server.scala [hunk 12]
```diff
@@ -17,15 +17,21 @@
...
+import org.http4s.parser.AdditionalRules
 import org.http4s.util.Writer
 
 object Server extends HeaderKey.Internal[Server] with HeaderKey.Singleton {
   def apply(id: ProductId): Server =
     new Server(id, Nil)
 
   override def parse(s: String): ParseResult[Server] =
-    HttpHeaderParser.SERVER(s)
+    ParseResult.fromParser(parser, "invalid Server")(s)
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/Server.scala:28]:
```suggestion
    ParseResult.fromParser(parser, "Invalid Server header")(s)
```

File: core/src/main/scala/org/http4s/headers/User-Agent.scala [hunk 13]
```diff
@@ -17,15 +17,21 @@
...
+import org.http4s.parser.AdditionalRules
 import org.http4s.util.Writer
 
 object `User-Agent` extends HeaderKey.Internal[`User-Agent`] with HeaderKey.Singleton {
   def apply(id: ProductId): `User-Agent` =
     new `User-Agent`(id, Nil)
 
   override def parse(s: String): ParseResult[`User-Agent`] =
-    HttpHeaderParser.USER_AGENT(s)
+    ParseResult.fromParser(parser, "Invalid User-Agent")(s)
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/User-Agent.scala:28]:
```suggestion
    ParseResult.fromParser(parser, "Invalid User-Agent header")(s)
```

File: core/src/main/scala/org/http4s/headers/X-Forwarded-For.scala [hunk 14]
```diff
@@ -18,13 +18,20 @@ package org.http4s
...
+import cats.parse._
 import java.net.InetAddress
-import org.http4s.parser.HttpHeaderParser
+import org.http4s.internal.parsing.Rfc7230
 import org.http4s.util.Writer
 
 object `X-Forwarded-For` extends HeaderKey.Internal[`X-Forwarded-For`] with HeaderKey.Recurring {
   override def parse(s: String): ParseResult[`X-Forwarded-For`] =
-    HttpHeaderParser.X_FORWARDED_FOR(s)
+    ParseResult.fromParser(parser, "Invalid X-Forwarded-For")(s)
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/X-Forwarded-For.scala:28]:
```suggestion
    ParseResult.fromParser(parser, "Invalid X-Forwarded-For header")(s)
```

File: core/src/main/scala/org/http4s/headers/X-Forwarded-For.scala [hunk 15]
```diff
@@ -18,13 +18,20 @@ package org.http4s
...
+  private[http4s] val parser: Parser[`X-Forwarded-For`] =
+    Rfc7230
+      .headerRep1(
+        (Uri.Ipv4Address.parser.map(_.toInet4Address).backtrack | Uri.Ipv6Address.parser.map(
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/X-Forwarded-For.scala:32]: This one can't be a hostname?

File: core/src/main/scala/org/http4s/internal/parsing/Rfc7230.scala [hunk 16]
```diff
@@ -57,11 +57,26 @@ private[http4s] object Rfc7230 {
...
     b *> a <* b
+  private def between[A](a: Parser[Any], b: Parser0[A], c: Parser[Any]): Parser[A] =
+    a *> b <* c
 
   /*quoted-string  = DQUOTE *( qdtext / quoted-pair ) DQUOTE*/
   val quotedString: Parser[String] =
     surroundedBy(qdText.orElse(quotedPair).rep0.string, dquote)
 
+  /* HTAB / SP / %x21-27 / %x2A-5B / %x5D-7E / obs-text */
+  val CText =
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/internal/parsing/Rfc7230.scala:68]: Nit: all the other parsers are lowercase names.

File: core/src/main/scala/org/http4s/parser/AdditionalRules.scala [hunk 17]
```diff
@@ -221,4 +221,10 @@ private[http4s] object AdditionalRules {
       case _: NumberFormatException => None
     }
   }
+
+  val ServerAgentParser: P[(ProductId, List[ProductIdOrComment])] = {
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/parser/AdditionalRules.scala:225]: Can we get this on the companion of `ProductIdOrComment`?  Putting models in this package gets us circular dependencies.

**Review by rossabaker — COMMENTED** [review]
Very nice work.  The suggested edits are just me being a consistency fanatic.  A couple bigger things to think about, but this is great overall.

File: core/src/main/scala/org/http4s/headers/Access-Control-Allow-Credentials.scala [hunk 18]
```diff
@@ -17,13 +17,15 @@
...
+import cats.parse.Parser
 import org.http4s.util.Writer
 
 object `Access-Control-Allow-Credentials`
     extends HeaderKey.Internal[`Access-Control-Allow-Credentials`] {
   override def parse(s: String): ParseResult[`Access-Control-Allow-Credentials`] =
-    HttpHeaderParser.ACCESS_CONTROL_ALLOW_CREDENTIALS(s)
+    ParseResult.fromParser(parser, "invalid Access-Control-Allow-Credentials")(s)
+
+  private[http4s] val parser = Parser.string("true").as(`Access-Control-Allow-Credentials`())
```
[the remark this answers is not in the document]
Comment (m-sp) [reply core/src/main/scala/org/http4s/headers/Access-Control-Allow-Credentials.scala:28]: yeah that one is weird. seems it also is case sensitive which was unexpected https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials

File: core/src/main/scala/org/http4s/headers/X-Forwarded-For.scala (hunk 15 above)
Comment (m-sp) [reply core/src/main/scala/org/http4s/headers/X-Forwarded-For.scala:32]: it's not defined in any spec but according to the mdn it's just ip addresses: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For

File: core/src/main/scala/org/http4s/headers/Host.scala (hunk 7 above)
Comment (m-sp) [reply core/src/main/scala/org/http4s/headers/Host.scala:33]: just to be sure that I under stand RFC7230. we could just delegate to `Uri.Host.parser` now? and remove the comment

File: core/src/main/scala/org/http4s/parser/AdditionalRules.scala (hunk 17 above)
Comment (m-sp) [reply core/src/main/scala/org/http4s/parser/AdditionalRules.scala]: https://github.com/http4s/http4s/pull/4208/commits/c6a368b9b9a64f5744880ec3628ed806e93e3f5a

File: core/src/main/scala/org/http4s/internal/parsing/Rfc7230.scala (hunk 16 above)
Comment (m-sp) [reply core/src/main/scala/org/http4s/internal/parsing/Rfc7230.scala]: https://github.com/http4s/http4s/pull/4208/commits/c6a368b9b9a64f5744880ec3628ed806e93e3f5a

File: core/src/main/scala/org/http4s/headers/Host.scala (hunk 7 above)
Comment (m-sp) [reply core/src/main/scala/org/http4s/headers/Host.scala]: https://github.com/http4s/http4s/pull/4208/commits/190d6766dc60f05351db722b8ae03ba4b15765de

**Review by hamnis — APPROVED** [review]

File: core/src/main/scala/org/http4s/headers/Access-Control-Allow-Credentials.scala (hunk 18 above)
[the remark this answers is not in the document]
Comment (rossabaker) [reply core/src/main/scala/org/http4s/headers/Access-Control-Allow-Credentials.scala:28]: Yeah, I was about to call that out, and found my way to the same page. :laughing: 

The obvious answer on the model is to make the header a case object, but we use the companion object for the keys.  We could keep a cached value of it.

**Comment by rossabaker** [comment]
Great job!
